LSPS4/README.md

```diff
@@ -0,0 +1,1923 @@
+  timer for the PPP has passed and the LSP has failed the
+  underlying PPP.
+  The error `data` object contains the required field
+  `unknown_ppp_ids`, which is an array of PPP identifiers
+  containing the PPPs which were not found (and were
+  presumably timed out).
+  * The client SHOULD remove the indicated PPPs from its
+    own copy of the PPP Map, and should then re-evaluate
+    if it can accept the PPPs (for instance, it might
+    have received a `lsps4.you_have_new_ppp` while this
+    call was being handled by the LSP) in the PPP Map.
+* `insufficient_liquidity` (3) - The client has insufficient
+  inbound liquidity to actually receive the entire set of
+  `ppp_ids`, and none of the `ppp_ids` were forwarded.
+  * The client SHOULD continue to the next step (5 and 6) of
+    the flow.
+
+If the `lsps4.forward_ppps` call succeeds, the LSP SHOULD
+make a best-effort attempt to deliver all the specified
+PPPs.
+
+* For HTLC-type PPPs:
+  * The LSP SHOULD send an `update_add_htlc` to the client,
+    via any channel that has sufficient capacity to the
+    client.
+
+As noted above, the LSP SHOULD make a best-effort attempt to
+deliver all the specified PPPs.
+However, race conditions can cause the actual inbound
+liquidity to change from when it was checked in the atomic
+section, to when the LSP is now sending `update_add_htlc`s
+to the client:
+
+* The channel initiator might send an `update_fees` which
+  increases the onchain feerate, such that an HTLC that would
+  have barely fit in the channel, no longer fits since the
+  increased onchain feerate prices it out of the channel.
+* Some automated system in the client or LSP (such as the
+  one checking that the `update_fees` from the remote are
+  reasonable) might close a channel that the LSP was planning
+  to send an HTLC through.
+* The client might be trying to mess with the LSP and
+  broadcasts a currently-unrevoked commitment transaction,
+  closing the channel before the LSP can deliver an HTLC it
+  was planning to send via that channel.
+* The LSP might allow specific channels to be indicated via
+  their SCID, and an unrelated payment might go through a
+  channel that the LSP was intending to send one of the
+  `lsps4.forward_ppps` HTLCs through.
+* One party might have sent `shutdown` for a channel that
+  was intended to be used for forwarding.
+
+In such a case the LSP, for each PPP it was unable to handle
+normally due to a race condition that changes the liquidity
+towards the client:
+
+* MUST do one of the following:
+  * (preferred) add the PPP back to the PPP Map under a
+    new PPP identifier and resumes its hold timer.
+    It re-sends `lsps4.you_have_new_ppp` under the new PPP
+    identifier.
+    If its hold timer has already expired at this point, the
+    LSP MUST NOT do this, but MUST do the alternative below
+    instead.
+  * (unpreferred) NOT add the PPP back to the PPP Map and
+    fail the payment with `temporary_channel_failure`.
```

Comment:
These things are hard to accomplish if you're building an LSP on top of CLN or LND. You would forward the htlc through the htlc interceptor interface, and you don't get a second chance. If the channel is not there, it results in unknown_next_peer. Not much you can do about it. (I see this is mostly addressed in the rationale section below)